### PR TITLE
VarChar

### LINKED ISSUE #176: char and varchar conversion
It should probably be possible to convert `Text` and `String` to and from `char` and `varchar`.

### PR DESCRIPTION
**Pull request by ilyakooo0**
resolves #176 

Didn't have the opportunity to test this whatsoever.

### CODE REVIEW

File: squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs [hunk 1]
```diff
@@ -831,6 +845,20 @@ newtype Only x = Only { fromOnly :: x }
 instance Generic (Only x)
 instance HasDatatypeInfo (Only x)
 
+-- | A refined text type for use with 'varchar' and 'char'.
+-- The constructor is not exposed. You have to use @constructVarChar@ and @unVarChar@.
+newtype VarChar (n :: Nat) = VarChar Strict.Text
+  deriving (Eq,Ord,Read,Show)
+
+constructVarChar :: forall n. KnownNat n => Strict.Text -> Maybe (VarChar n)
```
Comment (echatav) [inline squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs]: Can you rename this to `varChar`?

File: squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs [hunk 2]
```diff
@@ -831,6 +845,20 @@ newtype Only x = Only { fromOnly :: x }
...
+constructVarChar t =
+  if Strict.Text.length t <= (fromIntegral $ natVal @n Proxy)
+    then Just . VarChar $ t
+    else Nothing
+
+unVarChar :: VarChar n -> Strict.Text
```
Comment (echatav) [inline squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs]: Can you rename this to `getVarChar`?

File: squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs [hunk 3]
```diff
@@ -831,6 +845,20 @@ newtype Only x = Only { fromOnly :: x }
 instance Generic (Only x)
 instance HasDatatypeInfo (Only x)
 
+-- | A refined text type for use with 'varchar' and 'char'.
+-- The constructor is not exposed. You have to use @constructVarChar@ and @unVarChar@.
+newtype VarChar (n :: Nat) = VarChar Strict.Text
```
Comment (echatav) [inline squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs]: Can you move this all to the `PG` module next to other storage newtypes and add a `type instance PG (VarChar n) = 'PGvarchar n`?

File: squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs [hunk 4]
```diff
@@ -594,6 +600,14 @@ instance FromValue 'PGtext Strict.Text where fromValue = Decoding.text_strict
 instance FromValue 'PGtext Lazy.Text where fromValue = Decoding.text_lazy
 instance FromValue 'PGtext String where
   fromValue = Strict.Text.unpack <$> Decoding.text_strict
+instance KnownNat n => FromValue ('PGvarchar n) (VarChar n) where
+  fromValue = (constructVarChar <$> Decoding.text_strict) >>= \case
+      Just t -> return t
+      Nothing -> failure "Could not construct VarChar."
+instance KnownNat n => FromValue ('PGchar n) (VarChar n) where
```
Comment (echatav) [inline squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs]: I'm not sure this is right. `CHAR(n)` are fixed length, you'd need another newtype to describe them.

Comment (ilyakooo0) [reply squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs]: As far as I know, `char` would just get padded with spaces if the length is smaller.

It depends on whether we actually want that.

Comment (ilyakooo0) [reply squeal-postgresql/src/Squeal/PostgreSQL/Binary.hs]: I think it would actually be better to force the exact length.

**Pull request merged**